src/lib/io/sol-iio.c

```diff
@@ -595,22 +615,25 @@ channel_get_pure_name(struct sol_iio_channel *channel)
             channel_pure_name = calloc(1, channel_name_len + 1);
             original_channel_pure_name = channel_pure_name;
             for (i = 0; i < channel_name_len; i++) {
-                if (isalpha((uint8_t)channel->name[i]) || channel->name[i] == '-' || channel->name[i] == '_')
-                    *channel_pure_name++ = channel->name[i];
-                else if (isdigit((uint8_t)channel->name[i])) {
+                if (isalpha((uint8_t)channel_name[i]) || channel_name[i] == '-'
+                    || channel_name[i] == '_')
+                    *channel_pure_name++ = channel_name[i];
+                else if (isdigit((uint8_t)channel_name[i])) {
                     modified = true;
                     continue;
                 }
             }
-            if (modified)
+            if (modified) {
+                free(internsity_channel_name);
                 return original_channel_pure_name;
+            }
 
             free(original_channel_pure_name);
-            return NULL;
+            return internsity_channel_name;
         }
     }
 
-    return NULL;
+    return internsity_channel_name;
```

Comment:
I guess that this is over complicated...
A simple approach to use would be (highlevel pseudocode):
  - copy `channel->name` to a `channel_name[NAME_MAX]` buffer
  - `strendswith(channel_name, "_both") || strendswith(channel_name, "_ir")` ->strip from channel_name, doing `channel_name[strlen(channel_name) - 5 (or 3)] = '\0'`
  - let current code act, but on `channel_name` instead of `channel->name`
On last step, you could actually do the same for `_x`, `_y` and `_z` stripping, so code would be uniform, but it isn't a must for now.
Any thoughts?